pkg/libdevfile/uris.go

```diff
@@ -0,0 +1,123 @@
+package libdevfile
+
+import (
+	"errors"
+	"net/url"
+	"sort"
+	"strings"
+
+	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/api/v2/pkg/attributes"
+	"github.com/devfile/api/v2/pkg/validation"
+	"github.com/devfile/library/pkg/devfile/parser"
+	"github.com/devfile/library/pkg/devfile/parser/data/v2/common"
+)
+
+const _importSourceAttributeUriPrefix = "uri: "
+
+// GetReferencedLocalFiles returns the local files referenced by the Devfile. This includes:
+// - the non-inlined Kubernetes and Openshift components
+// - the Dockerfiles of Image components
+// - the parent devfile
+// - resursively, the local files referenced by the parent Devfile
+// The passed Devfile must be flattened
+func GetReferencedLocalFiles(devfileObj parser.DevfileObj) (result []string, err error) {
+
+	setResult := map[string]struct{}{}
+
+	parent := devfileObj.Data.GetParent()
+	if parent != nil {
+		return nil, errors.New("devfile must be flattened")
+	}
+
+	components, err := devfileObj.Data.GetComponents(common.DevfileOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, component := range components {
+		var componentType v1alpha2.ComponentType
+		componentType, err = common.GetComponentType(component)
+		if err != nil {
+			return nil, err
+		}
+
+		switch componentType {
+		case v1alpha2.KubernetesComponentType:
+			setResult, err = appendUriIfFile(setResult, component.Kubernetes.Uri)
+			if err != nil {
+				return nil, err
+			}
+
+		case v1alpha2.OpenshiftComponentType:
+			setResult, err = appendUriIfFile(setResult, component.Openshift.Uri)
+			if err != nil {
+				return nil, err
+			}
+
+		case v1alpha2.ImageComponentType:
+			if component.Image.Dockerfile != nil {
+				setResult, err = appendUriIfFile(setResult, component.Image.Dockerfile.Uri)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+
+		setResult, err = getFromAttributes(setResult, component.Attributes)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	commands, err := devfileObj.Data.GetCommands(common.DevfileOptions{})
+	if err != nil {
+		return nil, err
+	}
+	for _, command := range commands {
+		setResult, err = getFromAttributes(setResult, command.Attributes)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	result = make([]string, 0, len(setResult))
+	for k := range setResult {
+		result = append(result, k)
+	}
+	sort.Strings(result)
+	return result, nil
+}
+
+// appendUriIfFile appends uri to the result if the uri is a local path
+func appendUriIfFile(result map[string]struct{}, uri string) (map[string]struct{}, error) {
+	if uri != "" {
+		u, err := url.Parse(uri)
+		if err != nil {
+			return nil, err
+		}
+		if u.Scheme == "" {
```

Comment:
Just curious - is the `file` URI scheme something possible here to reference local files? Or should we always expect files relative to the devfile.yaml? The Devfile spec is pretty vague about this:
> Location in a file fetched from a uri.

Reply:
I'm pretty sure it is related to the directory containing the devfile. For example, the doc for parent URI is:

```
URI Reference of a parent devfile YAML file. It can be a full URL or a relative URI with the current devfile as the base URI.
```